core/src/nl/tudelft/watchdog/core/logic/ui/listeners/staticanalysis/CheckStyleChecksMessagesFetcher.java

```diff
@@ -0,0 +1,127 @@
+package nl.tudelft.watchdog.core.logic.ui.listeners.staticanalysis;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Properties;
+
+import nl.tudelft.watchdog.core.util.WatchDogLogger;
+import nl.tudelft.watchdog.core.util.WatchDogUtilsBase;
+
+/**
+ * Hello and welcome to this class full of reflection, hacks and other shenanigans.
```

Comment:
While it is written in a comical way, the story is full of technical details that are required to understand what is going on. Therefore I can't shorten this bit :cry:

Reply:
I had a go at it, based heavily on what you wrote. You would only need to fill in two questions. Thanks

```
* Warning: This class is full of classloader fixes, reflection, hacks to support Checkstyle warnings in IntelliJ and Eclipse. There are Checkstyle IDE plugins (for Eclipse and IntelliJ) and the plain Checkstyle jar 
* itself. The main complication comes from IntelliJ depending on Checkstyle via dynamically loading its jar over a {@link ClassLoader}. In Eclipse, we do not know whether the user has installed the Checkstyle plugin at all, 
* which means we have to dynamically load the Checkstyle classes, too.
*
* We need multiple classloaders to load the correct files 

which?

* The first classloader is created in the CheckStyle plugin. This plugin is configured
* to load several jars, including the CheckStyle jar. This loader therefore has access to CheckStyle
* classes. In the code below, this loader is called "checkStylePluginClassLoader".
*
* Besides the actual classes, other configuration resources from Checkstyle must also be loaded. This configuration however partly
* resides in CheckStyle jar. Other files reside in the plugin. To that end, the IntelliJ plugin implements a so-called
* "csaccess" layer. This layer has its own classloader and has access to: * 1. the csaccess files as implemented by the plugin and 2. the resources available in CheckStyle.
* This loader is named "checkStylePackageLoaderClassLoader" in the code below.
*
* The overall implementation design is therefore as follows: Load the configuration and other plugin-relevant
* classes in our current classloader. Once we got the relevant classes, obtain the classloader created by
* the plugin and load the
* CheckStyle checks (as defined in "com.puppycrawl.tools.checkstyle.checks"). Once we got all checks,
* we obtain their corresponding messages.
*
* Since these messages are stored in resources named "messages.properties" and based on the packagename,
* we need to use the last classloader to fetch the resources from there. We process the resources in
* {@link #addMessagesToCheckStyleBundleForLoadedChecks(ClassLoader, Map, ClassLoader)}.
* Observe that this method takes 2 classloaders, as we need both the classes (to obtain the resource path)
* and the other one 

which other one?

* to obtain the actual resource.
*
* If you need to update this code, think carefully about which classloader can load what. When we originally wrote this class, it was mostly based on
* {@see https://github.com/jshiell/checkstyle-idea/blob/master/src/csaccess/java/org/infernus/idea/checkstyle/service/cmd/OpCreateChecker.java}
* This class contained important documentation about its usage of "moduleClassLoader" and the "loaderOfCheckedCode".
```